### PR TITLE
BIGTOP-3873: Add OpenEuler 22.03 support

### PR DESCRIPTION
**Pull request by MacChen02**
### Description of PR
The PR support the OpenEuler 22.03, include building puppet and slaves docker env for compile and smoke test etc.

### How was this patch tested?
I confirmed that some of components were successfully compiled on OpenEuler 22.03, including ZK, oozie, hive etc.

### For code changes:

- [ ] Does the title or this PR starts with the corresponding JIRA issue id (BIGTOP-3876)
- [ ] Make sure that newly added files do not have any licensing issues. When in doubt refer to https://www.apache.org/licenses/

### CODE REVIEW

**Review by guyuqi — COMMENTED** [review]
Really thanks for working on it.
There are several points needed to be clarified:
OpenEuler support would not be added to the next comming 3.2 release.
We will test it on Arm and  x86 platform including the related smoke tests. 
More discussion would be launched if other community guys show their concerns for this PR.

File: docker/bigtop-puppet/build.sh [hunk 1]
```diff
@@ -63,5 +64,10 @@ case "${OS}-${VERSION}" in
     ;;
 esac
 
-docker build -t bigtop/puppet:${PREFIX}-${OS}-${VERSION}${ARCH} .
+if [ $OS == "openeuler/openeuler" ];then
+  docker build -t ${PREFIX}-${OS}-${VERSION}${ARCH} .
```
Comment (guyuqi) [inline docker/bigtop-puppet/build.sh:68]: The build command should be : `./build.sh trunk-openeuler-20.3`,
not be `./build.sh trunk-openeuler/openeuler-22.03`

Comment (MacChen02) [reply docker/bigtop-puppet/build.sh:68]: @guyuqi  thx, it will be modified.

Comment (MacChen02) [reply docker/bigtop-puppet/build.sh]: @guyuqi The build command was modifed, the commit:
https://github.com/apache/bigtop/pull/1051/commits/a46e50e432b830907deca32bfdbaf63dabbb1ee7

Now, the build command is :  **./build.sh trunk-openeuler-20.3**

Comment (docularxu) [reply docker/bigtop-puppet/build.sh]: > Now, the build command is : **./build.sh trunk-openeuler-20.3**

you mean "22.03", right?

Comment (MacChen02) [reply docker/bigtop-puppet/build.sh]: @docularxu yes

**Comment by kahohwang** [comment]
My OS Version is openEuler 22.03LTS SP1, and after I ran puppetize.sh, i got an error by checking puppet version:

**Comment by kahohwang** [comment]
My OS version is 22.03LTS SP1, and i got an error after I ran puppetize.sh:
**[root@bigtopc1h1 bin]# puppet -v
/usr/local/share/gems/gems/puppet-7.20.0/lib/puppet/thread_local.rb:6:in `<top (required)>': uninitialized constant Concurrent::RubyThreadLocalVar (NameError)
Did you mean?  Concurrent::RubyThreadPoolExecutor
	from /usr/local/share/gems/gems/puppet-7.20.0/lib/puppet/context.rb:1:in `require_relative'
	from /usr/local/share/gems/gems/puppet-7.20.0/lib/puppet/context.rb:1:in `<top (required)>'
	from /usr/local/share/gems/gems/puppet-7.20.0/lib/puppet.rb:44:in `require_relative'
	from /usr/local/share/gems/gems/puppet-7.20.0/lib/puppet.rb:44:in `<module:Puppet>'
	from /usr/local/share/gems/gems/puppet-7.20.0/lib/puppet.rb:41:in `<top (required)>'
	from /usr/local/share/gems/gems/puppet-7.20.0/lib/puppet/util/command_line.rb:12:in `require_relative'
	from /usr/local/share/gems/gems/puppet-7.20.0/lib/puppet/util/command_line.rb:12:in `<top (required)>'
	from <internal:/usr/share/rubygems/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from <internal:/usr/share/rubygems/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from /usr/local/share/gems/gems/puppet-7.20.0/bin/puppet:4:in `<top (required)>'
	from /usr/local/bin/puppet:25:in `load'
	from /usr/local/bin/puppet:25:in `<main>'**

after I updated the puppet version from 7.20 to 7.23(the latest version), issue solved.  And I check the puppetize.sh, it  wrote as:

**# openEuler ruby version is 3.X,so use puppet-7.20.0.
         gem install puppet -v 7.20.0**

I don't know if I could modify the script directly so I mark it here.

**Comment by MacChen02** [comment]
@kahohwang 
Please take the newest version PR, the version of puppet has been upgraded from 7.20.0 to 7.22.0 last month.
image

**Comment by kahohwang** [comment]
> @kahohwang Please take the newest version PR, the version of puppet has been upgraded from 7.20.0 to 7.22.0 last month. image
> 
> Our verification work is carried out on puppet-7.22.0 .

Sorry, I don't know why I clone from the branch last night and I still got the old version.🤣  I will try the new one. thx.

**Comment by qingzhongli** [comment]
OpenEuler 22.03 does not provide the `R` and `pandoc` required to packaging Spark, is it necessary to improve the logic about OpenEuler in `bigtop_toolchain/manifests/renv.pp`?
```
[root@bigtop-dev ~]# docker run --rm -t openeuler/openeuler:22.03 bash -c "yum install -y R pandoc"
OS                                                                                                    260 kB/s | 3.4 MB     00:13    
everything                                                                                            672 kB/s |  16 MB     00:25    
EPOL                                                                                                  285 kB/s | 3.8 MB     00:13    
debuginfo                                                                                             291 kB/s | 3.9 MB     00:13    
source                                                                                                149 kB/s | 1.7 MB     00:11    
update                                                                                                638 kB/s |  14 MB     00:23    
update-source                                                                                          23 kB/s | 247 kB     00:10    
No match for argument: R
No match for argument: pandoc
Error: Unable to find a match: R pandoc
[root@bigtop-dev ~]# 
```

**Comment by guyuqi** [comment]
> OpenEuler 22.03 does not provide the `R` and `pandoc` required to packaging Spark, is it necessary to improve the logic about OpenEuler in `bigtop_toolchain/manifests/renv.pp`?
> 
> ```
> [root@bigtop-dev ~]# docker run --rm -t openeuler/openeuler:22.03 bash -c "yum install -y R pandoc"
> OS                                                                                                    260 kB/s | 3.4 MB     00:13    
> everything                                                                                            672 kB/s |  16 MB     00:25    
> EPOL                                                                                                  285 kB/s | 3.8 MB     00:13    
> debuginfo                                                                                             291 kB/s | 3.9 MB     00:13    
> source                                                                                                149 kB/s | 1.7 MB     00:11    
> update                                                                                                638 kB/s |  14 MB     00:23    
> update-source                                                                                          23 kB/s | 247 kB     00:10    
> No match for argument: R
> No match for argument: pandoc
> Error: Unable to find a match: R pandoc
> [root@bigtop-dev ~]# 
> ```

It seems this PR has updated the `renv.pp` and offer `R` and `pandoc` for openEuler by @MacChen02

**Comment by MacChen02** [comment]
> Sorry for the late reply，the R language has been changed to the source code installation method. @guyuqi

**Comment by guyuqi** [comment]
> > Sorry for the late reply，the R language has been changed to the source code installation method. @guyuqi

That's ok. It seems there are some conflicts in this PR and it need to be rebased on `master`.

**Comment by MacChen02** [comment]
> > > Sorry for the late reply，the R language has been changed to the source code installation method. @guyuqi
> 
> That's ok. It seems there are some conflicts in this PR and it need to be rebased on `master`.

The code conflict resolved, please help to check it again，thks.

File: docker/bigtop-puppet/build.sh [hunk 2]
```diff
@@ -23,7 +23,6 @@ if [ $# != 1 ]; then
   echo "Usage: build.sh <PREIX-OS-VERSION>"
   echo
   echo "Example: build.sh trunk-centos-7"
-  echo "openeuler: Example: build.sh trunk-openeuler/openeuler-22.03"
```
Comment (docularxu) [inline docker/bigtop-puppet/build.sh]: This line (and the below lines L67-71) was added by you in your [previous commit in the same PR](https://github.com/apache/bigtop/pull/1051/commits/da5e75519d2110abfcb27fb118eb43affcb14720).
That seems unnecessary for keeping the history clean.

Would you please squash the two commits?

Comment (MacChen02) [reply docker/bigtop-puppet/build.sh]: The example output information deleted.

File: docker/bigtop-puppet/build.sh [hunk 3]
```diff
@@ -63,5 +63,10 @@ case "${OS}-${VERSION}" in
     ;;
 esac
 
+if [ $OS == "openeuler" ];then
+  OPENEULER_OS="${OS}/${OS}"
+  sed -i -e "s|${OS}:${VERSION}|$OPENEULER_OS:${VERSION}|" ./Dockerfile
+fi
+
```
Comment (docularxu) [inline docker/bigtop-puppet/build.sh:70]: Suggest to use the 'case'/'esac' logic, as set by in L54~L64. Like this:
```
  openeuler-*)
  OPENEULER_OS="${OS}/${OS}"
      sed -i -e "s|${OS}:${VERSION}|$OPENEULER_OS:${VERSION}|" ./Dockerfile
    ;;

```

File: docker/bigtop-puppet/build.sh (hunk 2 above)
Comment (docularxu) [reply docker/bigtop-puppet/build.sh]: No. I mean you submitted two commits which modified the same file, but one added several lines, the other deleted them. Why don't you put these changes into one commit?

This adds: https://github.com/apache/bigtop/pull/1051/commits/da5e75519d2110abfcb27fb118eb43affcb14720 
This deletes: https://github.com/apache/bigtop/pull/1051/commits/a46e50e432b830907deca32bfdbaf63dabbb1ee7

Comment (MacChen02) [reply docker/bigtop-puppet/build.sh]: These modifications were introduced to resolve different issues during the validation process.

File: docker/bigtop-puppet/build.sh (hunk 3 above)
Comment (MacChen02) [reply docker/bigtop-puppet/build.sh]: tks, it's done. 
[67721e5](https://github.com/apache/bigtop/pull/1051/commits/67721e505533cb54a023aa5f2a3828966daff52d)

**Comment by guyuqi** [comment]
@MacChen02

Thanks for your continued efforts on this PR.

Iwasaki is working on v3.2.1 release. 
To avoid some conflicts, I've created the branch: `openEuler-support`.
What about rebasing this PR to branch `openEuler-support `?
I'll merge this PR after the rebasing.
Thanks again.

**Comment by MacChen02** [comment]
> @MacChen02
> 
> Thanks for your continued efforts on this PR.
> 
> Iwasaki is working on v3.2.1 release. To avoid some conflicts, I've created the branch: `openEuler-support`. What about rebasing this PR to branch `openEuler-support `? I'll merge this PR after the rebasing. Thanks again.

Agreed with your suggestion, tks.

**Comment by guyuqi** [comment]
Duplicated to https://github.com/apache/bigtop/pull/1127

**Pull request closed without merging**